internal/webhook/external/artifactory.go

```diff
@@ -0,0 +1,196 @@
+		},
+	)
+}
+
+// artifactoryWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from Jfrog Artifactory.
+type artifactoryWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newArtifactoryWebhookReceiver returns a new instance of artifactoryWebhookReceiver.
+func newArtifactoryWebhookReceiver(
+	c client.Client,
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &artifactoryWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Artifactory.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (a *artifactoryWebhookReceiver) getReceiverType() string {
+	return artifactory
+}
+
+// getSecretValues implements WebhookReceiver.
+func (a *artifactoryWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[artifactorySecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for a Artifactory WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (a *artifactoryWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+
+		token, ok := a.secretData[artifactorySecretDataKey]
+		if !ok {
+			xhttp.WriteErrorJSON(w, nil)
+			return
+		}
+
+		sig := r.Header.Get("x-jfrog-event-auth")
+		if sig == "" {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New("missing signature"),
+					http.StatusUnauthorized,
+				),
+			)
+			return
+		}
+
+		mac := hmac.New(sha256.New, token)
+		mac.Write(requestBody)
+		computedSig := hex.EncodeToString(mac.Sum(nil))
+
+		if !hmac.Equal([]byte(sig), []byte(computedSig)) {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New("unauthorized"),
+					http.StatusUnauthorized,
+				),
+			)
+			return
+		}
+
+		logger.Info("Received Artifactory webhook",
+			"raw event", string(requestBody),
+		)
+
+		var payload artifactoryEvent
+		if err := json.Unmarshal(requestBody, &payload); err != nil {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(errors.New("invalid request body"),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		switch payload.EventType {
+		case "pushed":
+			parsed, err := url.Parse(payload.Origin)
+			if err != nil {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(
+						fmt.Errorf("failed to parse origin URL: %w", err),
+						http.StatusBadRequest,
+					),
+				)
+				return
+			}
+			repoURL, err := url.JoinPath(
+				parsed.Hostname(),
+				payload.Data.RepoKey,
+				payload.Data.ImageName,
+			)
+			if err != nil {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(
+						fmt.Errorf("failed to construct repository URL: %w", err),
+						http.StatusBadRequest,
+					),
+				)
```

Comment:
This shouldn't happen and if it somehow does, I think we ought to count this as an internal error. To have gotten this far, the request was good.